src/forest.jl

```diff
@@ -0,0 +1,52 @@
+mutable struct Forest{T<:Integer}
+    counter::T
+    intmap::Dict{Tuple{T,T},T}
+    parents::Vector{T}
+    ranks::Vector{T}
+    ntrees::T
+end
+
+function Forest{T}(n::Integer) where {T<:Integer}
+    counter = zero(T)
+    intmap = Dict{Tuple{T,T},T}()
+    sizehint!(intmap, n)
+    parents = collect(Base.OneTo(T(n)))
+    ranks = zeros(T, T(n))
+    ntrees = zero(T)
+    return Forest{T}(counter, intmap, parents, ranks, ntrees)
+end
+
+function Base.push!(forest::Forest{T}, edge::Tuple{T,T}) where {T<:Integer}
+    forest.counter += 1
+    forest.intmap[edge] = forest.counter
+    forest.ntrees += one(T)
+    return edge
+end
+
+function _find_root!(parents::Vector{T}, index_edge::T) where {T<:Integer}
+    @inbounds p = parents[index_edge]
+    @inbounds if parents[p] != p
```

Comment:
```suggestion
    p = parents[index_edge]
    if parents[p] != p
```
`@inbounds` is almost never a good idea unless we prove it is essential

Reply:
Can you explain why?
It's how was implemented `DisjointSet` in `DataStructures.jl`.
For performance reasons, I don't see why we should not use it.

Reply:
Older Julia libraries like DataStructures.jl had a much more liberal use of `@inbounds` than modern code bases. As the Julia compiler becomes smarter, it becomes less obvious where the macro actually speeds things up (see this [Discourse thread](https://discourse.julialang.org/t/when-does-inbounds-increase-performance/123827)).
But more importantly, misusing `@inbounds` can have very bad correctness implications or straight up crash the REPL (I've seen it with SparseDiffTools.jl for example). Here, your uses of `@inbounds` are unchecked, meaning that there is no guarantee for `index_edge` to be within the keys of `parents`.
I am very very very very strongly in favor of never using `@inbounds`. If you come up with an example where bounds checking alone takes more than 10% of the time, then yes we can reconsider (after adding proper checks), but it will also mean we have overcome every other complexity hurdle in our way. Until then, the risk is not worth the reward.